controllers/cluster_controller_test.go

```diff
@@ -211,4 +219,173 @@ var _ = Describe("Cluster Reconciler", func() {
+			if err := k8sClient.Get(ctx, key, machine); err != nil {
+				return false
+			}
+			return len(machine.Finalizers) > 0
+		}, timeout).Should(BeTrue())
+
+		// patch machine noderef
+		patchHelper, err := patch.NewHelper(machine, k8sClient)
+		Expect(err).NotTo(HaveOccurred())
+
+		machine.Status.NodeRef = &v1.ObjectReference{Kind: "Node", Name: "test-node"}
+		Expect(patchHelper.Patch(ctx, machine)).Should(BeNil())
+
+		// Assertion
+		key = client.ObjectKey{Name: cluster.Name, Namespace: cluster.Namespace}
+		Eventually(func() bool {
+			if err := k8sClient.Get(ctx, key, cluster); err != nil {
+				return false
+			}
+			return cluster.Status.ControlPlaneInitialized
+		}, timeout).Should(BeTrue())
+	})
 })
+
+func TestClusterReconciler_machineToCluster(t *testing.T) {
+	cluster := &clusterv1.Cluster{
+		TypeMeta: metav1.TypeMeta{
+			Kind: "Cluster",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-cluster",
+			Namespace: "test",
+		},
+		Spec:   clusterv1.ClusterSpec{},
+		Status: clusterv1.ClusterStatus{},
+	}
+
+	controlPlaneWithNoderef := &clusterv1.Machine{
+		TypeMeta: metav1.TypeMeta{
+			Kind: "Machine",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "controlplaneWithnoderef",
+			Labels: map[string]string{
+				clusterv1.MachineClusterLabelName:      cluster.Name,
+				clusterv1.MachineControlPlaneLabelName: cluster.Name,
+			},
+		},
+		Status: clusterv1.MachineStatus{
+			NodeRef: &v1.ObjectReference{
+				Kind:      "Node",
+				Namespace: "test-node",
+			},
+		},
+	}
+	controlPlaneWithoutNoderef := &clusterv1.Machine{
+		TypeMeta: metav1.TypeMeta{
+			Kind: "Machine",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "controlplaneWithNilNoderef",
+			Labels: map[string]string{
+				clusterv1.MachineClusterLabelName:      cluster.Name,
+				clusterv1.MachineControlPlaneLabelName: cluster.Name,
+			},
+		},
+	}
+	nonControlPlane := &clusterv1.Machine{
+		TypeMeta: metav1.TypeMeta{
+			Kind: "Machine",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "Nocontrolplane",
+			Labels: map[string]string{
+				clusterv1.MachineClusterLabelName: cluster.Name,
+			},
+		},
+	}
+
+	tests := []struct {
+		name string
+		o    handler.MapObject
+		want []ctrl.Request
+	}{
+		{
+			name: "controlplane machine, noderef is set, should return cluster",
+			o: handler.MapObject{
+				Meta:   controlPlaneWithNoderef.GetObjectMeta(),
+				Object: controlPlaneWithNoderef,
+			},
+			want: []ctrl.Request{
+				{NamespacedName: client.ObjectKey{
+					Name:      cluster.Name,
+					Namespace: cluster.Namespace,
+				}},
+			},
+		},
+		{
+			name: "controlplane machine, noderef is not set",
+			o: handler.MapObject{
+				Meta:   controlPlaneWithoutNoderef.GetObjectMeta(),
+				Object: controlPlaneWithoutNoderef,
+			},
+			want: nil,
+		},
+		{
+			name: "not controlplane machine",
+			o: handler.MapObject{
+				Meta:   nonControlPlane.GetObjectMeta(),
+				Object: nonControlPlane,
+			},
+			want: nil,
+		},
```

Comment:
Do you also want to add a test case for "not controlplane machine, noderef is set"?